File: src/storeFx/goToPrevSong.ts
import * as Fx from '../fx/fx';
import type { Store } from '../store/types';
import * as StoreFx__Play from './play';
import { setCurrentTime } from './setCurrentTime';
import * as StoreFx__StartPlaying from './startPlaying';

export function goToPrevSong() {
  return Fx.bind(Fx.ask<{ store: Store }>(), ({ store }) => {
    const { player } = store.getState();

    if (player.currentTime > 3) {
      return Fx.bind(
        setCurrentTime(() => 0),
        () => StoreFx__Play.make(),
      );
    }

    const queued = player.queuedSongIds;
    const len = queued.length;

    const currentIndex =
      player.currentSongId == null ? 0 : queued.indexOf(player.currentSongId);

    return StoreFx__StartPlaying.make(queued[(currentIndex + len - 1) % len]);
  });
}

File: src/storeFx/goToPrevSong.ml
open Fx
open Monad_syntax

let make () =
  let* Deps.{ store } = ask () in
  let state = store |. Zustand.getState in
  if state.player.currentTime > 3.0 then
    let* () = SetCurrentTime.make (Fun.const 0.0) in
    Play.make ()
  else
    let queued = state.player.queuedSongIds in
    let len = queued |> Js.Array.length
    and currentIndex =
      state.player.currentSongId
      |> Option.map (fun value -> queued |> Js.Array.indexOf ~value)
      |> Option.value ~default:0
    in
    let current =
      queued |. Belt.Array.getUnsafe ((len + currentIndex - 1) mod len)
    in
    StartPlaying.make ~current ()
